### PR TITLE
Adds ESM support

### PR DESCRIPTION
**Pull request by benjaminsehl**
Updates build script, ESLint, default modules, and module packages.

Builds and looks right … but still some tests failing. Would love help to understand what's going on… a bit stumped here.

### CODE REVIEW

File: packages/lexical/Lexical.esm.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+import Lexical from './dist/Lexical.esm.js';
+
+export default Lexical;
```
Comment (TrySound) [inline packages/lexical/Lexical.esm.js:11]: This breaks all named imports

File: packages/lexical-code/package.json [hunk 2]
```diff
@@ -9,7 +9,12 @@
   ],
   "license": "MIT",
   "version": "0.12.0",
-  "main": "LexicalCode.js",
+  "main": "LexicalCode.cjs.js",
+  "module": "LexicalCode.esm.js",
```
[the remark this answers is not in the document]
Comment (benjaminsehl) [reply packages/lexical-code/package.json]: I've got with stating both `.mjs` and `.cjs` to be explicit — let me know if it's better to keep `.cjs` as simply `.js`.

**Comment by benjaminsehl** [comment]
Thanks for the feedback @TrySound — I believe I've addressed what you mentioned. Still can't figure out why those 4 unit tests aren't working.

*(Disclaimer: I'm just a PM that can code a bit … but a total novice when it comes to build tooling, though I'm liking learning, thanks again!)*

**Comment by ivailop7** [comment]
@benjaminsehl and @TrySound any luck on this one?

**Comment by TrySound** [comment]
It seems to me named exports problem still not fixed. Now won't even work with bundlers.
[My approach does](https://github.com/facebook/lexical/pull/4784) not bring esm bundles though at least let use named imports in node native esm.

Plus dual packages with cjs and esm support often have problems with different tools. So maybe postpone dual mode and ship esm only later.

**Comment by TrySound** [comment]
The most important thing is to get rid of webpack4 which is sadly still used a lot.
In node part enable type:module to write tests for custom lexical plugins and deal with other related issues (remix still not work with type:module).

**Comment by benjaminsehl** [comment]
I'm going to close this out in favour of yours @TrySound

**Pull request closed without merging**